### ISSUE TITLE
onDelay function

### ISSUE THREAD

**Issue by vanderlee** [bgrins/bindWithDelay#1]
new jQuery replaced .bind() and a few others with .on, so I wanted a version that worked like the .on() function.

Here's my full onDelay() (I removed the "With" in the name for my own use, as I thought it was redundant). It's closely based on your code. If you want to include it in your repository, go ahead. No author notice required.

---

(function($) {
$.fn.onDelay = function(events, selector, data, handler, timeout, throttle) {
    if ($.isFunction(selector)) {
        throttle    = handler;
        timeout     = data;
        handler     = selector;
        data        = undefined;
        selector    = undefined;
    } else if ($.isFunction(data)) {
        throttle    = timeout;
        timeout     = handler;
        handler     = data;
        data        = undefined;
    }

```
// Allow delayed function to be removed with handler in unbind function
handler.guid = handler.guid || ($.guid && $.guid++);

// Bind each separately so that each element has its own delay
return this.each(function() {
    var wait = null;

    function callback() {
        var event = $.extend(true, { }, arguments[0]);
        var that = this;
        var throttler = function() {
            wait = null;
            handler.apply(that, [event]);
        };

        if (!throttle) { clearTimeout(wait); wait = null; }
        if (!wait) { wait = setTimeout(throttler, timeout); }
    }

    callback.guid = handler.guid;

    $(this).on(events, selector, data, callback);
});
```

}
})(jQuery);

**Comment by bgrins** [comment]
Cool, thanks for sharing!  I put together a little demo to test it out: http://jsfiddle.net/bgrins/KT4bF/

**Comment by LucaRosaldi** [comment]
Now it works with deferreds too! http://jsfiddle.net/LucaRosaldi/KT4bF/1/

bgrins, you didn't update the code on the master branch, though.

**Comment by vanderlee** [comment]
bgrins, Will you be updating your code to include LucaRisoldi's version of .onDelay()?
If not, I will contact LucaRosaldi and create a new project for it so it doesn't get lost.

**Comment by bgrins** [comment]
@vanderlee @LucaRosaldi sorry about that, this fell off my radar.  I will take a look at it today and see if I can add support for the `on` function using this code as part of the library.  Are you thinking this should just be a separate file in master?

**Comment by vanderlee** [comment]
I'd probably go the same way as jQuery itself; implement just the `onDelay` method and make `bindDelay` a proxy for `onDelay`. Similarly, you could also include a `liveDelay` as a proxy to `onDelay`, though there is not a lot value in doing so.

**Comment by bgrins** [comment]
That is a good point about `onDelay` and `bindWithDelay`.  I'm not sure we even need `bindWithDelay` anymore, as `on` is now the preferred method of binding.  If it is a quick one liner to proxy the bind function over I guess it wouldn't hurt to keep it there.

Is this the new proposed API?
- `onDelay`: matches http://api.jquery.com/on/ with two additional parameters - a time and an optional boolean indicating whether it should be throttled.
- `bindDelay`: matches http://api.jquery.com/bind/ with two additional parameters - a time and an optional boolean indicating whether it should be throttled.

**Comment by vanderlee** [comment]
I'd prefer to have only `onDelay` as the "preferred method of binding" and `bindWithDelay` for backwards compatibility for current users.
`bindDelay` seems redundant; nobody has used it before, and new users should use `onDelay`. On the other hand; it's probably a simple proxy to include it.

**Comment by bgrins** [comment]
OK, so the plugin file will expose `onDelay` and `bindWithDelay`, though if possible we should make it so that `bindWithDelay` will be easy to include / exclude as it is mostly there for backwards compatibility at this point.

**Comment by bgrins** [comment]
I've added a new file for now here: https://github.com/bgrins/bindWithDelay/blob/master/onDelay.js.  It is not working on the demo page, because it seems to get messed up when `data` is passed, like this:

```
$(window).bindWithDelay("resize", {when:'delay'}, updateResize, 1000);
```

Need to review the code and probably write a few tests to ensure the API is handled properly.  It would actually be great to pull in qunit and copy some of the functionality of the standard jQuery events test: https://github.com/jquery/jquery/blob/master/test/unit/event.js.  

At least the code is in the project now - I will wait to close the issue and mark the file as ready until we have confirmed it is matching the `on` API in all cases.
